### ISSUE TITLE
bot.cancelAllDialogs() called inside a dialog handler causes a crash

### ISSUE THREAD

**Issue by etiennellipse** [howdyai/botkit#1808]
I am trying to leave all the existing dialogs from within an handler; checking if the user really wishes to leave, then I call `await bot.cancelAllDialogs()` in the handler.

I am getting the following error at runtime:
```
(node:76242) UnhandledPromiseRejectionWarning: TypeError: Cannot read property 'state' of undefined
    at BotkitConversation.<anonymous> (/Users/etienne/Code/botkit-example/node_modules/botkit/lib/conversation.js:617:62)
    at Generator.next (<anonymous>)
    at /Users/etienne/Code/botkit-example/node_modules/botkit/lib/conversation.js:7:71
    at new Promise (<anonymous>)
    at __awaiter (/Users/etienne/Code/botkit-example/node_modules/botkit/lib/conversation.js:3:12)
    at BotkitConversation.end (/Users/etienne/Code/botkit-example/node_modules/botkit/lib/conversation.js:614:16)
    at BotkitConversation.<anonymous> (/Users/etienne/Code/botkit-example/node_modules/botkit/lib/conversation.js:558:35)
    at Generator.next (<anonymous>)
    at fulfilled (/Users/etienne/Code/botkit-example/node_modules/botkit/lib/conversation.js:4:58)
    at process._tickCallback (internal/process/next_tick.js:68:7)
```

After some investigation, I found that the end() method in the BotkitConversation class uses `dc.activeDialog` which has been nulled by cancelAllDialogs(). Hence, the crash.

Here is the code example for reproduction:
```
const { BotkitConversation } = require('botkit');

module.exports = function(controller) {
    const DIALOG_SUBJECT_ID = 'subject-dialog';
    const DIALOG_INTERRUPT_ID = 'interrupt';

    const subjectDialog = new BotkitConversation(DIALOG_SUBJECT_ID, controller);

    for (let i = 0; i < 30; i++) {
        subjectDialog.ask(`Message #${i}`, [], `conv-${i}`);
    }

    controller.addDialog(subjectDialog);

    const interruptDialog = new BotkitConversation(DIALOG_INTERRUPT_ID, controller);

    interruptDialog.ask({
        text: "This subject is not interesting to you?",
        quick_replies: [
            {
                title: 'Sure',
                payload: 'Sure'
            },
            {
                title: 'Nope',
                payload: 'Nope'
            }
        ]
    }, [{
        pattern: 'Sure',
        handler: async function (response, convo, bot) {
            await bot.say('Let\'s continue then!');
        },
    },{
        pattern: 'Nope',
        handler: async function (response, convo, bot) {
            await bot.say('Let\'s leave then!');
            await bot.cancelAllDialogs();
        },
    }]);

    controller.addDialog(interruptDialog);

    controller.hears('go', ['message', 'direct_message'], async (bot, message) => {
        await bot.beginDialog(DIALOG_SUBJECT_ID);
    });

    controller.hears('not interested', ['message','direct_message'], async (bot, message) => {
        await bot.beginDialog(DIALOG_INTERRUPT_ID);
    });
};

```

### Context:

- Botkit version: 4.5.0
- Messaging Platform: Web 
- Node version: v10.16.3
- Os: MacOS

**Comment by chahat-arora** [comment]
@etiennellipse hey, did you find any solution or workaround for this?

**Comment by etiennellipse** [comment]
It works from a `bot.interrupts` handler, though. It would seem that it fails because the handler runs inside a dialog, maybe?

```
        controller.interrupts(async () => true, ["hello"], async (bot: CerebroBotWorker) => {
            await bot.cancelAllDialogs();
            await bot.say("Hey, you are back! Wanna talk?");
        });
```

**Comment by benbrown** [comment]
should be fixed in v4.6

**Comment by chahat-arora** [comment]
I don't think it's fixed in `v4.6`.

**Comment by benbrown** [comment]
@chahat-arora can you  please send some sample code that exhibits this issue? I am not able to reproduce it.
